contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -5851,3 +5864,112 @@ is_bbf_db_ddladmin_operation(Oid namespaceId)
 
 	return false;
 }
+
+static bool
+allow_storing_init_privs(Oid objoid, Oid classoid, int objsubid)
+{
+	ObjectAddress			address;
+	Oid						nspoid = InvalidOid;
+	ObjectType				objtype;
+	PltsqlInitPrivsOptions	init_privs_opt = ERROR_INIT_PRIVS;
+
+	/*
+	 * Check if it is create/upgrade script of babelfishpg_tsql extension
+	 * Otherwise return true -- Allow storing.
+	 */
+	if (!(creating_extension &&
+		OidIsValid(CurrentExtensionObject) &&
+		CurrentExtensionObject == get_extension_oid("babelfishpg_tsql", true)))
+		return true;
+
+	/* 
+	 * There are 3 category of handling
+	 * 1. SAVE_INIT_PRIVS    : Check if it is objects created during CREATE extension and
+	 *                         store initial privs for them. system, information_schema_tsql
+	 *                         objects and pltsql language are the examples of it.
+	 * 
+	 * 2. DISCARD_INIT_PRIVS : If it is schema contained object within system created
+	 *                         TSQL schema like master, msdb or tempdb OR user created schema,
+	 *                         Do not store initial privileges for them.
+	 * 
+	 * 3. ERROR_INIT_PRIVS   : The default case when above 2 conditions doesn't match then error
+	 *                         out. To avoid error please classify it between above 2 condtions. 
+	 */
+	ObjectAddressSet(address, classoid, objoid);
+	objtype = get_object_type(classoid, objoid);
+	if (objtype == OBJECT_SCHEMA)
+	{
+		nspoid = objoid;
+	}
+	else
+	{
+		nspoid = get_object_namespace(&address);
+	}
+	if (OidIsValid(nspoid)) /* Schema contained objects */
```

Comment:
for all the calls below, what are perf implications? I guess it would be OK if we come in this code path for create/alter extension only

Reply:
I think there won't be much perf. implication:
1. This code path get executed only in case of CREATE/ALTER extension.
2. Specifically for GRANT/REVOKE statements only.